File: CSharp/SpaceVIL/SpaceVIL/Windows/WindowLayoutBox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;

namespace SpaceVIL
{
    static class WindowLayoutBox
    {
        static public Dictionary<string, WindowLayout> windows = new Dictionary<string, WindowLayout>();
        static public int ActiveWindow;
        static public void InitWindow(WindowLayout _layout)
        {
            windows.Add(_layout.GetWindowName(), _layout);
            ItemsLayoutBox.InitLayout(_layout.Id);

            //add filling frame
            //ALL ATTRIBUTES STRICTLY NEEDED!!!
            _layout.SetWindow(new Frame());
            _layout.GetWindow().SetHandler(_layout);
            _layout.GetWindow().SetItemName(_layout.GetWindowName());
            _layout.GetWindow().SetWidth(_layout.GetWidth());
            _layout.GetWindow().SetHeight(_layout.GetHeight());
            _layout.GetWindow().SetWidthPolicy(SizePolicy.Expand);
            _layout.GetWindow().SetHeightPolicy(SizePolicy.Expand);

            ItemsLayoutBox.AddItem(_layout, _layout.GetWindow());
        }
        static public WindowLayout GetWindowInstance(string name)
        {
            if (windows.ContainsKey(name))
                return windows[name];
            else return null;
        }
        static public void WindowDispatcher(int sender_id)
        {
            ActiveWindow = sender_id;
            foreach (var wnd in windows)
            {
                if (wnd.Value.IsOutsideClickClosable && (wnd.Value.Id != ActiveWindow))
                    wnd.Value.Close();
            }
        }
        static public string[] GetListOfWindows() => windows.Keys.ToArray();
        static public void PrintStoredWindows()
        {
            foreach (var item in windows.Keys.ToArray())
            {
                Console.WriteLine(item);
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spacevil/WindowLayoutBox.java
package com.spacevil;

import com.spacevil.Flags.LayoutType;
import com.spacevil.Flags.SizePolicy;

import java.util.*;
import java.util.stream.*;

public final class WindowLayoutBox {
    static Map<String, WindowLayout> windowsName = new HashMap<>();
    static Map<UUID, WindowLayout> windowsUUID = new HashMap<>();
    static List<WindowsPair> currentCallingPair = new LinkedList<>();
    static WindowLayout lastFocusedWindow;
    //static private Object locker = new Object();

    static void initWindow(WindowLayout _layout) {
        windowsName.put(_layout.getWindowName(), _layout);
        windowsUUID.put(_layout.getId(), _layout);

        ItemsLayoutBox.initLayout(_layout.getId());

        // add filling frame
        // ALL ATTRIBUTES STRICTLY NEEDED!!!
        WContainer container = new WContainer();
        container.setHandler(_layout);
        // System.out.println(_layout.getWindowName());
        container.setItemName(_layout.getWindowName());
        container.setWidth(_layout.getWidth());
        container.setMinWidth(_layout.getMinWidth());
        container.setMaxWidth(_layout.getMaxWidth());
        container.setHeight(_layout.getHeight());
        container.setMinHeight(_layout.getMinHeight());
        container.setMaxHeight(_layout.getMaxHeight());
        container.setWidthPolicy(SizePolicy.EXPAND);
        container.setHeightPolicy(SizePolicy.EXPAND);

        _layout.setWindow(container);
        ItemsLayoutBox.addItem(_layout, container, LayoutType.STATIC);
    }

    static void removeWindow(WindowLayout _layout) {
        windowsName.remove(_layout.getWindowName());
        windowsUUID.remove(_layout.getId());
    }

    /**
     * Try to show WindowLayout object using its UUID
     * @return if showing successful
     */
    static public boolean tryShow(UUID guid) {
        WindowLayout wnd = WindowLayoutBox.getWindowInstance(guid);
        if (wnd != null) {
            wnd.show();
            return true;
        }
        return false;
    }

    /**
     * Try to show WindowLayout object using its name
     * @return if showing successful
     */
    static public boolean tryShow(String name) {
        WindowLayout wnd = WindowLayoutBox.getWindowInstance(name);
        if (wnd != null) {
            wnd.show();
            return true;
        }
        return false;
    }

    /**
     * @return WidowLayout object by its name
     */
    static public WindowLayout getWindowInstance(String name) {
        if (windowsName.containsKey(name))
            return windowsName.get(name);
        else
            return null;
    }

    /**
     * @return WidowLayout object by its UUID
     */
    static public WindowLayout getWindowInstance(UUID guid) {
        if (windowsUUID.containsKey(guid))
            return windowsUUID.get(guid);
        else
            return null;
    }

    static void addToWindowDispatcher(WindowLayout sender_wnd) {
        WindowsPair pair = new WindowsPair();
        pair.WINDOW = sender_wnd;
        if (lastFocusedWindow == null) {
            pair.GUID = sender_wnd.getId();// root
            lastFocusedWindow = sender_wnd;
        } else
            pair.GUID = lastFocusedWindow.getId();
        currentCallingPair.add(pair);
    }

    static void setCurrentFocusedWindow(WindowLayout wnd) {
        lastFocusedWindow = wnd;
    }

    static void setFocusedWindow(CoreWindow window) {
        window.getHandler().setFocus(true);
    }

    static void removeFromWindowDispatcher(WindowLayout sender_wnd) {
        List<WindowsPair> pairs_to_delete = new LinkedList<WindowsPair>();
        for (WindowsPair windows_pair : currentCallingPair) {
            if (windows_pair.WINDOW.equals(sender_wnd)) {
                pairs_to_delete.add(windows_pair);
            }
        }

        for (WindowsPair pairs : pairs_to_delete) {
            currentCallingPair.remove(pairs);
        }

        pairs_to_delete = null;
    }

    /**
     * @return list of all windows names
     */
    static public String[] getListOfWindows() {
        String[] result = new String[windowsName.size()];
        windowsName.entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toList()).toArray(result);
        return result;
    }

    /**
     * Print all windows names
     */
    static public void printStoredWindows() {
        for (String item : getListOfWindows()) {
            System.out.println(item);
        }
    }
}